azure/scope/machine.go

```diff
@@ -257,21 +253,62 @@ func (m *MachineScope) NICSpecs() []azure.ResourceSpecGetter {
 		}
 	}
 
-	// If NAT gateway is not enabled and node has no public IP, then the NIC needs to reference the LB to get outbound traffic.
+	if m.Role() == infrav1.Node && m.AzureMachine.Spec.AllocatePublicIP {
+		spec.PublicIPName = azure.GenerateNodePublicIPName(m.Name())
+	}
+	// If the NAT gateway is not enabled and node has no public IP, then the NIC needs to reference the LB to get outbound traffic.
 	if m.Role() == infrav1.Node && !m.Subnet().IsNatGatewayEnabled() && !m.AzureMachine.Spec.AllocatePublicIP {
 		spec.PublicLBName = m.OutboundLBName(m.Role())
 		spec.PublicLBAddressPoolName = m.OutboundPoolName(m.OutboundLBName(m.Role()))
 	}
 
-	if m.Role() == infrav1.Node && m.AzureMachine.Spec.AllocatePublicIP {
-		spec.PublicIPName = azure.GenerateNodePublicIPName(m.Name())
+	if len(m.AzureMachine.Spec.NetworkInterfaces) < 1 {
+		// Retains for backwards compatibility with the old behavior, prior to the addition of configurable AzureNetworkInterfaces.
+		return spec
 	}
 
-	if m.cache != nil {
-		spec.SKU = &m.cache.VMSKU
+	// We always use the first network interface as the primary network interface.
+	primaryNICIndex := 0
+	nic := m.AzureMachine.Spec.NetworkInterfaces[primaryNICIndex]
```

Comment:
we shouldn't rely on the position of the item in the list API field here, lists are not strictly ordered and there is nothing that guarantees the order won't change through conversions, users making changes, etc.

Reply:
The `Primary: true` field in the ARM Compute API as I understand it does not enforce interface attachment order. Only the order of the config in the array does that. My understanding of the `Primary: true` field is that it indicates the interface that is used for traffic to the IMDS and maybe also `168.63.129.16`. I have observed that if you attempt to access the IMDS over an interface that is not designated as `Primary: true`, the IMDS is inaccessible.

Slices in Go and arrays in JSON are strictly ordered. If this were not the case this problem would already exist in many places. Ordinality in the list in the call to the ARM Compute API is also the only way to signal to the Azure platform which order the network interfaces should be attached to the virtual hardware so that the OS finds them in the expected order.

The same approach is taken in [terraform](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_virtual_machine#network_interface_ids).

Similar is done [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/ca53d743a546c8849a1fde92cdfb39ce558aa35f/api/v1beta2/awsmachine_types.go#L134)

This makes for a very natural UX, because as the user configures their network interfaces, they naturally assume the first in the list is going to be the first in the Azure Compute API, and `eth0` in the OS.

On a slight tangent, it is theoretically possible that a CAPZ user may have a use case to set `primary: true` on an interface that is not the first in the list, but that seems very unlikely, and I believe it exposes an unnecessary foot-gun for the function of things like IMDS.

Reply:
> The Primary: true field in the ARM Compute API as I understand it does not enforce interface attachment order. Only the order of the config in the array does that

this part is surprising to me, why would the API have a field `Primary` if that gets ignored? do you have a link to docs or code that states this? 

> Slices in Go and arrays in JSON are strictly ordered

this is true, but here I'm more concerned about the user experience of assuming order, and for round trip conversion. I thought I recalled something about k8s API conventions and not relying on order for fields of type array, but I may be misremembering. @sbueringer @vincepri do you know anything about this? Also are there any ClusterClass implications?

> Similar is done [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/ca53d743a546c8849a1fde92cdfb39ce558aa35f/api/v1beta2/awsmachine_types.go#L134)

I can't see anywhere in CAPA code where we are making any assumptions about the item at position 0 being different than the others, AFAICT it's just performing the same attach operation for every item in the list https://cs.github.com/kubernetes-sigs/cluster-api-provider-aws/blob/ca53d743a546c8849a1fde92cdfb39ce558aa35f/pkg/cloud/services/ec2/instances.go?q=.NetworkInterfaces#L246

Here it's different because we're not just passing through the list to Azure, we're configuring the NICs differently based on their position in the list.

> This makes for a very natural UX, because as the user configures their network interfaces, they naturally assume the first in the list is going to be the first in the Azure Compute API, and eth0 in the OS.

This is my main concern, we shouldn't make any implicit assumptions when it comes to UX, if we make an assumption about order being significant then it should be clearly documented in the docs and API field description. This is different from other existing fields in the API, such as [subnets](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/types.go#L140) and [security rules](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/types.go#L242), where order does not matter.

Reply:
The order should be stable. K8s core relies on stable order of lists too for things such as [executing initContainers in order](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#detailed-behavior). Conversion logic looks correct to me to preserve order, relying on json marshal/unmarshal. 

Argo workflows CRD uses list order for ordered operations as well. 

The Primary field in the Compute Vm API isn't unused but it doesn't set interface index. I just looked it up and it is used to set influence the default route in the OS (I suspect by modifying the DHCP response for the VM). It is documented [here](https://learn.microsoft.com/en-us/azure/virtual-network/virtual-network-network-interface-vm#constraints).

Reply:
Missed the mention somehow

> this is true, but here I'm more concerned about the user experience of assuming order, and for round trip conversion. I thought I recalled something about k8s API conventions and not relying on order for fields of type array, but I may be misremembering. @sbueringer @vincepri do you know anything about this? Also are there any ClusterClass implications?

I'm not aware of conventions around not relying on array ordering. Not sure if that answers the question but ClusterClass does preserve array ordering. So if you have a new array/slice field and that's passed through ClusterClass or set via a variable the array order will be preserved.